Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp

```diff
@@ -369,317 +411,385 @@ ProcessCode VertexPerformanceWriter::writeT(
-            }
-
-            break;
-          }
-        }
+    std::vector<std::pair<SimVertexBarcode, double>> contributingTruthVertices;
 
-        if (!foundMatchingParticle) {
-          ACTS_DEBUG("Track has no matching truth particle.");
-        }
+    double totalTrackWeight = 0;
+    for (const Acts::TrackAtVertex& trk : tracksAtVtx) {
+      if (trk.trackWeight < m_cfg.minTrkWeight) {
+        continue;
       }
 
-      if (tracksAtVtx.size() != trackIndices.size()) {
-        ACTS_ERROR("Not all tracks at vertex have a matching input track.");
+      totalTrackWeight += trk.trackWeight;
+
+      std::optional<ConstTrackProxy> trackOpt = findTrack(trk);
+      if (!trackOpt.has_value()) {
+        ACTS_DEBUG("Track has no matching input track.");
+        continue;
       }
-    } else {
-      for (const SimParticle& particle : particles) {
-        contributingTruthVertices.emplace_back(particle.particleId().vertexId(),
-                                               1);
+      const ConstTrackProxy& inputTrk = *trackOpt;
+      std::optional<SimParticle> particleOpt = findParticle(inputTrk);
+      if (!particleOpt.has_value()) {
+        ACTS_VERBOSE("Track has no matching truth particle.");
+      } else {
+        contributingTruthVertices.emplace_back(
+            SimBarcode(particleOpt->particleId()).vertexId(), trk.trackWeight);
       }
     }
 
-    double recoVertexTrackWeights = 0;
-    for (const Acts::TrackAtVertex& trk : tracksAtVtx) {
-      recoVertexTrackWeights += trk.trackWeight;
-    }
-
     // Find true vertex that contributes most to the reconstructed vertex
-    std::map<SimVertexBarcode, double> truthVertexWeights;
+    std::map<SimVertexBarcode, std::pair<int, double>> fmap;
     for (const auto& [vtxId, weight] : contributingTruthVertices) {
-      truthVertexWeights[vtxId] += weight;
-    }
-    auto truthVertexMatch = std::max_element(
-        truthVertexWeights.begin(), truthVertexWeights.end(),
-        [](const auto& a, const auto& b) { return a.second < b.second; });
-
-    if (truthVertexMatch == truthVertexWeights.end()) {
-      ACTS_DEBUG("No truth vertex found for reconstructed vertex.");
-      continue;
+      ++fmap[vtxId].first;
+      fmap[vtxId].second += weight;
     }
-    SimVertexBarcode truthVertexId = truthVertexMatch->first;
-    double truthVertexTrackWeights = truthVertexMatch->second;
-
-    // Count number of reconstructible tracks on truth vertex
-    std::uint32_t nTracksOnTruthVertex = 0u;
-    for (const SimParticle& particle : selectedParticles) {
-      SimBarcode vtxId =
-          SimBarcode(particle.particleId()).setParticle(0).setSubParticle(0);
-      if (vtxId == truthVertexId) {
-        ++nTracksOnTruthVertex;
+    double maxOccurrence = -1;
+    SimVertexBarcode maxOccurrenceId = -1;
+    for (const auto& [vtxId, counter] : fmap) {
+      if (counter.second > maxOccurrence) {
+        maxOccurrenceId = vtxId;
+        maxOccurrence = counter.second;
```

Comment:
@andiwand I am confused here. 
MaxOccurrence here is the max of the contributing vtx weight, right? Since counter.second is the vtx weight, not a counter.

Reply:
Right the name is not great. I will change this in another PR